receipt of non-float number data will cause a panic in newResult function

The newResult function of driver.go will always fail if the reading is something other than a string or float64.  This is because the JSON unmarshaling will always interpret any number as a float64 (see https://golang.org/pkg/encoding/json/#Unmarshal).  Therefore, the type of the reading in the newResult function for any number (whether int, uint, etc.) will always be a float64 and cause a panic in the checks.  Best I could do was just assume that the profileValueType for the data was going to be correct and cast to the appropriate type.

func newResult(deviceObject models.DeviceObject, ro models.ResourceOperation, reading interface{}) (*sdkModel.CommandValue, error) {
	var result = &sdkModel.CommandValue{}
	var err error
	var resTime = time.Now().UnixNano() / int64(time.Millisecond)
	var profileValueType = strings.ToLower(deviceObject.Properties.Value.Type)


	switch profileValueType {
	case "bool":
		result, err = sdkModel.NewBoolValue(&ro, resTime, reading.(bool))
	case "string":
		result = sdkModel.NewStringValue(&ro, resTime, reading.(string))
	case "uint8":
		result, err = sdkModel.NewUint8Value(&ro, resTime, uint8(reading.(float64)))
	case "uint16":
		result, err = sdkModel.NewUint16Value(&ro, resTime, uint16(reading.(float64)))
	case "uint32":
		result, err = sdkModel.NewUint32Value(&ro, resTime, uint32(reading.(float64)))
	case "uint64":
		result, err = sdkModel.NewUint64Value(&ro, resTime, uint64(reading.(float64)))
	case "int8":
		result, err = sdkModel.NewInt8Value(&ro, resTime, int8(reading.(float64)))
	case "int16":
		result, err = sdkModel.NewInt16Value(&ro, resTime, int16(reading.(float64)))
	case "int32":
		result, err = sdkModel.NewInt32Value(&ro, resTime, int32(reading.(float64)))
	case "int64":
		result, err = sdkModel.NewInt64Value(&ro, resTime, int64(reading.(float64)))
	case "float32":
		result, err = sdkModel.NewFloat32Value(&ro, resTime, reading.(float32))
	case "float64":
		result, err = sdkModel.NewFloat64Value(&ro, resTime, reading.(float64))
	default:
		err = fmt.Errorf("return result fail, none supported value type: %v", deviceObject.Properties.Value.Type)
	}

	return result, err
}

Yes, you're right. I reproduce this error by MQTT client tool. 

Simulate an async value with the terminal tool:
```
./mosquitto_pub -t DataTopic -m '{"name":"MQTT test device","cmd":"randIntNum","method":"get","randIntNum":123}'
```
It throw error:
```
panic: interface conversion: interface {} is float64, not int64

goroutine 67 [running]:
github.com/edgexfoundry/device-mqtt-go/internal/driver.checkValueInRange(0xc00015e1e0, 0x5, 0x14bd9a0, 0x7, 0x14fa9a0, 0xc00015e1c0, 0x14fa900)
	/Users/bruce/go/src/github.com/edgexfoundry/device-mqtt-go/internal/driver/driver.go:530 +0x1536
```
And I will work on this issue.

--- a/internal/driver/driver.go
+++ b/internal/driver/driver.go
@@ -300,7 +300,7 @@ func newResult(deviceObject models.DeviceObject, ro models.ResourceOperation, re
 	var err error
 	var resTime = time.Now().UnixNano() / int64(time.Millisecond)
 	var profileValueType = strings.ToLower(deviceObject.Properties.Value.Type)
-	var readingValueType = reflect.TypeOf(reading).String()
+	var readingValueType = strings.ToLower(reflect.TypeOf(reading).String())
 
 	if readingValueType == "int" {
 		reading = int64(reading.(int))
@@ -630,10 +630,48 @@ func convertReadingValueType(profileValueType string, readingValueType string, r
 		case "uint64":
 			reading = uint64(reading.(uint64))
 		}
-	} else if readingValueType == "float64" && profileValueType == "float32" {
-		reading = float32(reading.(float64))
-	} else if readingValueType == "float32" && profileValueType == "float64" {
-		reading = float64(reading.(float32))
+	} else if readingValueType == "float64" {
+		switch profileValueType {
+		case "int8":
+			reading = int8(reading.(float64))
+		case "int16":
+			reading = int16(reading.(float64))
+		case "int32":
+			reading = int32(reading.(float64))
+		case "int64":
+			reading = int64(reading.(float64))
+		case "uint8":
+			reading = uint8(reading.(float64))
+		case "uint16":
+			reading = uint16(reading.(float64))
+		case "uint32":
+			reading = uint32(reading.(float64))
+		case "uint64":
+			reading = uint64(reading.(float64))
+		case "float32":
+			reading = float32(reading.(float64))
+		}
+	} else if readingValueType == "float32" {
+		switch profileValueType {
+		case "int8":
+			reading = int8(reading.(float32))
+		case "int16":
+			reading = int16(reading.(float32))
+		case "int32":
+			reading = int32(reading.(float32))
+		case "int64":
+			reading = int64(reading.(float32))
+		case "uint8":
+			reading = uint8(reading.(float32))
+		case "uint16":
+			reading = uint16(reading.(float32))
+		case "uint32":
+			reading = uint32(reading.(float32))
+		case "uint64":
+			reading = uint64(reading.(float32))
+		case "float64":
+			reading = float64(reading.(float32))
+		}
 	}
 
 	return reading

--- a/internal/driver/driver.go
+++ b/internal/driver/driver.go
@@ -490,72 +490,102 @@ func handleReadingStringValue(profileValueType string, readingValueType string, 
 func checkValueInRange(profileValueType string, readingValueType string, reading interface{}) bool {
 	isValid := false
 
-	switch profileValueType {
-	case "string":
-		isValid = true
-	case "bool":
-		isValid = true
-	case "uint8":
-		val := reading.(uint64)
-		if val >= 0 && val <= math.MaxUint8 {
-			isValid = true
-		}
-	case "uint16":
-		val := reading.(uint64)
-		if val >= 0 && val <= math.MaxUint16 {
-			isValid = true
-		}
-	case "uint32":
-		val := reading.(uint64)
-		if val >= 0 && val <= math.MaxUint32 {
-			isValid = true
-		}
-	case "uint64":
+	if profileValueType == "string" || profileValueType == "bool" {
+		return true
+	}
+
+	if strings.Contains(profileValueType, "uint") {
 		var val uint64
 		if readingValueType == "int" {
 			val = uint64(reading.(int64))
-		} else {
+		} else if readingValueType == "uint" {
 			val = uint64(reading.(uint64))
-		}
-		maxiMum := uint64(math.MaxUint64)
-		if val >= 0 && val <= maxiMum {
-			isValid = true
-		}
-	case "int8":
-		val := reading.(int64)
-		if val >= math.MinInt8 && val <= math.MaxInt8 {
-			isValid = true
-		}
-	case "int16":
-		val := reading.(int64)
-		if val >= math.MinInt16 && val <= math.MaxInt16 {
-			isValid = true
-		}
-	case "int32":
-		val := reading.(int64)
-		if val >= math.MinInt32 && val <= math.MaxInt32 {
-			isValid = true
-		}
-	case "int64":
-		val := reading.(int64)
-		if val >= math.MinInt64 && val <= math.MaxInt64 {
-			isValid = true
-		}
-	case "float32":
-		var val float64
-		if readingValueType == "float32" {
-			val = (float64)(reading.(float32))
+		} else if readingValueType == "float32" {
+			val = uint64(reading.(float32))
 		} else {
-			val = reading.(float64)
+			val = uint64(reading.(float64))
 		}
-		if val >= math.SmallestNonzeroFloat32 && val <= math.MaxFloat32 {
-			isValid = true
+
+		switch profileValueType {
+		case "uint8":
+			if val >= 0 && val <= math.MaxUint8 {
+				isValid = true
+			}
+		case "uint16":
+			if val >= 0 && val <= math.MaxUint16 {
+				isValid = true
+			}
+		case "uint32":
+			if val >= 0 && val <= math.MaxUint32 {
+				isValid = true
+			}
+		case "uint64":
+			maxiMum := uint64(math.MaxUint64)
+			if val >= 0 && val <= maxiMum {
+				isValid = true
+			}
 		}
-	case "float64":
-		val := reading.(float64)
-		if val >= math.SmallestNonzeroFloat64 && val <= math.MaxFloat64 {
-			isValid = true
+		return isValid
+	}
+
+	if strings.Contains(profileValueType, "int") {
+		var val int64
+		if readingValueType == "int" {
+			val = int64(reading.(int64))
+		} else if readingValueType == "uint" {
+			val = int64(reading.(uint64))
+		} else if readingValueType == "float32" {
+			val = int64(reading.(float32))
+		} else {
+			val = int64(reading.(float64))
 		}
+
+		switch profileValueType {
+		case "int8":
+			if val >= math.MinInt8 && val <= math.MaxInt8 {
+				isValid = true
+			}
+		case "int16":
+			if val >= math.MinInt16 && val <= math.MaxInt16 {
+				isValid = true
+			}
+		case "int32":
+			if val >= math.MinInt32 && val <= math.MaxInt32 {
+				isValid = true
+			}
+		case "int64":
+			if val >= math.MinInt64 && val <= math.MaxInt64 {
+				isValid = true
+			}
+		}
+		return isValid
+	}
+
+	if strings.Contains(profileValueType, "float") {
+		var val float64
+		if readingValueType == "int" {
+			val = float64(reading.(int64))
+		} else if readingValueType == "uint" {
+			val = float64(reading.(uint64))
+		} else if readingValueType == "float32" {
+			val = float64(reading.(float32))
+		} else {
+			val = float64(reading.(float64))
+		}
+
+		switch profileValueType {
+		case "float32":
+			if val >= math.SmallestNonzeroFloat32 && val <= math.MaxFloat32 {
+				isValid = true
+			}
+		case "float64":
+			val := reading.(float64)
+			if val >= math.SmallestNonzeroFloat64 && val <= math.MaxFloat64 {
+				isValid = true
+			}
+		}
+
+		return isValid
 	}
 
 	return isValid

--- a/internal/driver/driver_test.go
+++ b/internal/driver/driver_test.go
@@ -1,6 +1,7 @@
 package driver
 
 import (
+	"fmt"
 	"testing"
 
 	"github.com/edgexfoundry/edgex-go/pkg/clients/logging"
@@ -24,22 +25,6 @@ func TestNewResult_bool(t *testing.T) {
 	cmdVal, _ := newResult(deviceObject, ro, reading)
 	val, err := cmdVal.BoolValue()
 	if val != true || err != nil {
-		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
-	}
-}
-
-func TestNewResult_string(t *testing.T) {
-	var reading interface{} = "test string"
-	deviceObject := models.DeviceObject{
-		Properties: models.ProfileProperty{
-			Value: models.PropertyValue{Type: "string"},
-		},
-	}
-	ro := models.ResourceOperation{}
-
-	cmdVal, _ := newResult(deviceObject, ro, reading)
-	val, err := cmdVal.StringValue()
-	if val != "test string" || err != nil {
 		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
 	}
 }
@@ -219,8 +204,120 @@ func TestNewResult_float64(t *testing.T) {
 	}
 }
 
-func TestNewResult_uint(t *testing.T) {
-	var reading interface{} = 123
+func TestNewResult_float64ToInt8(t *testing.T) {
+	var reading interface{} = float64(123.11)
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "int8"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.Int8Value()
+	if val != int8(reading.(float64)) || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_float64ToInt16(t *testing.T) {
+	var reading interface{} = float64(123.11)
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "int16"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.Int16Value()
+	if val != int16(reading.(float64)) || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_float64ToInt32(t *testing.T) {
+	var reading interface{} = float64(123.11)
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "int32"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.Int32Value()
+	if val != int32(reading.(float64)) || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_float64ToInt64(t *testing.T) {
+	var reading interface{} = float64(123.11)
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "int32"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.Int32Value()
+	if val != int32(reading.(float64)) || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_float64ToUint8(t *testing.T) {
+	var reading interface{} = float64(123.11)
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "uint8"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.Uint8Value()
+	if val != uint8(reading.(float64)) || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_float64ToUint16(t *testing.T) {
+	var reading interface{} = float64(123.11)
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "uint16"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.Uint16Value()
+	if val != uint16(reading.(float64)) || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_float64ToUint32(t *testing.T) {
+	var reading interface{} = float64(123.11)
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "uint32"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.Uint32Value()
+	if val != uint32(reading.(float64)) || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_float64ToUint64(t *testing.T) {
+	var reading interface{} = float64(123.11)
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{
 			Value: models.PropertyValue{Type: "uint64"},
@@ -230,13 +327,13 @@ func TestNewResult_uint(t *testing.T) {
 
 	cmdVal, _ := newResult(deviceObject, ro, reading)
 	val, err := cmdVal.Uint64Value()
-	if val != uint64(123) || err != nil {
+	if val != uint64(reading.(float64)) || err != nil {
 		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
 	}
 }
 
-func TestNewResult_float(t *testing.T) {
-	var reading interface{} = 123.0
+func TestNewResult_float64ToFloat32(t *testing.T) {
+	var reading interface{} = float64(123.11)
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{
 			Value: models.PropertyValue{Type: "float32"},
@@ -246,12 +343,44 @@ func TestNewResult_float(t *testing.T) {
 
 	cmdVal, _ := newResult(deviceObject, ro, reading)
 	val, err := cmdVal.Float32Value()
-	if val != float32(123) || err != nil {
+	if val != float32(reading.(float64)) || err != nil {
 		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
 	}
 }
 
-func TestNewResult_float32String(t *testing.T) {
+func TestNewResult_float64ToString(t *testing.T) {
+	var reading interface{} = float64(123.11)
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "string"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.StringValue()
+	if val != fmt.Sprintf("%v", reading) || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_string(t *testing.T) {
+	var reading interface{} = "test string"
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "string"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.StringValue()
+	if val != "test string" || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_stringToFloat32(t *testing.T) {
 	var reading interface{} = "123.0"
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{
@@ -267,7 +396,7 @@ func TestNewResult_float32String(t *testing.T) {
 	}
 }
 
-func TestNewResult_float64String(t *testing.T) {
+func TestNewResult_stringToFloat64(t *testing.T) {
 	var reading interface{} = "123.0"
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{
@@ -283,7 +412,7 @@ func TestNewResult_float64String(t *testing.T) {
 	}
 }
 
-func TestNewResult_int64String(t *testing.T) {
+func TestNewResult_stringToInt64(t *testing.T) {
 	var reading interface{} = "123"
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{
@@ -299,7 +428,7 @@ func TestNewResult_int64String(t *testing.T) {
 	}
 }
 
-func TestNewResult_int8String(t *testing.T) {
+func TestNewResult_stringToInt8(t *testing.T) {
 	var reading interface{} = "123"
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{
@@ -315,7 +444,7 @@ func TestNewResult_int8String(t *testing.T) {
 	}
 }
 
-func TestNewResult_uint8String(t *testing.T) {
+func TestNewResult_stringToUint8(t *testing.T) {
 	var reading interface{} = "123"
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{
@@ -331,7 +460,7 @@ func TestNewResult_uint8String(t *testing.T) {
 	}
 }
 
-func TestNewResult_uint32String(t *testing.T) {
+func TestNewResult_stringToUint32(t *testing.T) {
 	var reading interface{} = "123"
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{
@@ -347,7 +476,7 @@ func TestNewResult_uint32String(t *testing.T) {
 	}
 }
 
-func TestNewResult_uint64String(t *testing.T) {
+func TestNewResult_stringToUint64(t *testing.T) {
 	var reading interface{} = "123"
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{
@@ -363,7 +492,7 @@ func TestNewResult_uint64String(t *testing.T) {
 	}
 }
 
-func TestNewResult_boolString(t *testing.T) {
+func TestNewResult_stringToBool(t *testing.T) {
 	var reading interface{} = "true"
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{
@@ -379,7 +508,39 @@ func TestNewResult_boolString(t *testing.T) {
 	}
 }
 
-func TestNewResult_string2(t *testing.T) {
+func TestNewResult_numberToUint64(t *testing.T) {
+	var reading interface{} = 123
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "uint64"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.Uint64Value()
+	if val != uint64(123) || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_floatNumberToFloat32(t *testing.T) {
+	var reading interface{} = 123.0
+	deviceObject := models.DeviceObject{
+		Properties: models.ProfileProperty{
+			Value: models.PropertyValue{Type: "float32"},
+		},
+	}
+	ro := models.ResourceOperation{}
+
+	cmdVal, _ := newResult(deviceObject, ro, reading)
+	val, err := cmdVal.Float32Value()
+	if val != float32(123) || err != nil {
+		t.Errorf("Convert new result(%v) failed, error: %v", val, err)
+	}
+}
+
+func TestNewResult_numberToString(t *testing.T) {
 	var reading interface{} = 123
 	deviceObject := models.DeviceObject{
 		Properties: models.ProfileProperty{

